php-frontend/src/main/java/org/sonar/php/tree/symbols/SymbolVisitor.java

```diff
@@ -551,29 +549,50 @@ private void usageForNamespaceName(NamespaceNameTree namespaceName, Symbol.Kind
 
   @Override
   public void visitMemberAccess(MemberAccessTree tree) {
-    memberStatementDepth++;
-    if (memberStatementDepth >= MEMBER_ACCESS_MAXIMUM_DEPTH) {
-      if (LOG.isDebugEnabled()) {
-        LOG.debug("visitMemberAccess at \"{}...\" has reached the maximum depth of {}, switching to the next statement",
-          tree.toString().substring(0, MEMBER_ACCESS_LOG_TRUNCATION_LENGTH), MEMBER_ACCESS_MAXIMUM_DEPTH);
-      }
-      return;
+    boolean isCallChain = tree.object().is(Kind.FUNCTION_CALL);
+
+    if (isCallChain) {
+      // special handling of long call chains which happens in some DSL-like scenarios, including but not limited to SQL queries.
+      // Because we unwind the chain from the end, we don't know its exact length, and this branch will execute even for short chains.
+      Stream.iterate(tree, Objects::nonNull, (Tree next) -> {
+        if (next instanceof MemberAccessTree memberAccessTree) {
+          return memberAccessTree.object();
+        } else if (next instanceof FunctionCallTree functionCallTree) {
+          return functionCallTree.callee();
+        } else {
+          return null;
+        }
+      })
+        .forEach((Tree next) -> {
+          if (next instanceof MemberAccessTree memberAccessTree) {
+            visitMemberAccess(memberAccessTree, false);
+          } else if (next instanceof FunctionCallTree functionCallTree) {
+            visitFunctionCall(functionCallTree, false);
+          } else if (!next.is(Kind.NAMESPACE_NAME)) {
+            // namespace name is handled in a corresponding `visit*` method
+            next.accept(this);
+          }
+        });
```

Comment:
Although I like this use of `Stream.iterate(...)`, I'm not sure of the readability of the final result.
It's a bit odd to do the same `instanceof` checks twice, the usage of `iterate` and `forEach` visually split a logic that belong together.
```suggestion
      ExpressionTree current = tree;
      while (current != null) {
        if (current instanceof MemberAccessTree memberAccessTree) {
          visitMemberAccess(memberAccessTree, false);
          current = memberAccessTree.object();
        } else if (current instanceof FunctionCallTree functionCallTree) {
          visitFunctionCall(functionCallTree, false);
          current = functionCallTree.callee();
        } else {
          if (!current.is(Kind.NAMESPACE_NAME)) {
            current.accept(this);
          }
          current = null;
        }
      }
```
In the above code, we do each `instanceof` check once, which helps us follow the logic.
(those codes are not strictly equivalent)
I let you decide which form you prefer.